### PR TITLE
added kpm_update

### PR DESCRIPTION
**Pull request by octonawish-akcodes**
Tried to implement the kcl mod update similar to the metadata, Any guidance about how to add the integration with the LSP? #1031 

cc @Peefy @zong-zhe @He1pa

### CODE REVIEW

File: kclvm/driver/src/kpm_update.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                }
+                Err(err) => bail!("update failed with error: {}", err),
+            }
+        }
+        None => bail!("Manifest file '{}' not found in directory hierarchy", MANIFEST_FILE),
+    }
+}
+
+/// Get the path for the KCL executable.
+fn kcl() -> PathBuf {
```
Comment (zong-zhe) [inline kclvm/driver/src/kpm_update.rs:35]: This method is the same as https://github.com/kcl-lang/kcl/blob/7232ff1346a91d23bf97377f5a34a67aa188109f/kclvm/driver/src/kpm_metadata.rs#L140

You can move the method `kcl()` to file https://github.com/kcl-lang/kcl/blob/main/kclvm/driver/src/lib.rs.
And reuse it in both `kpm_metadata.rs` and `kpm_update.rs`

File: kclvm/driver/src/kpm_update.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    if lookup_in_path(executable_name) {
+        return executable_name.into();
+    }
+
+    executable_name.into()
+}
+
+/// [`lookup_in_path`] will search for an executable file [`exec`] in the environment variable ‘PATH’.
+///  If found, return true, otherwise return false.
+fn lookup_in_path(exec: &str) -> bool {
```
Comment (zong-zhe) [inline kclvm/driver/src/kpm_update.rs:54]: The same as the method kcl(), you can move it to https://github.com/kcl-lang/kcl/blob/main/kclvm/driver/src/lib.rs

File: kclvm/driver/src/kpm_update.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        .find_map(probe)
+        .is_some()
+}
+
+/// [`probe`] check if the given path points to a file.
+/// If it does, return [`Some`] of the path.
+/// If not, check if adding the current operating system's executable file extension (if any) to the path points to a file.
+/// If it does, return [`Some`] of the path with the extension added.
+/// If neither, return [`None`].
+fn probe(path: PathBuf) -> Option<PathBuf> {
```
Comment (zong-zhe) [inline kclvm/driver/src/kpm_update.rs:67]: The same as the method kcl(), you can move it to https://github.com/kcl-lang/kcl/blob/main/kclvm/driver/src/lib.rs

File: kclvm/driver/src/kpm_update.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        it => Some(path.with_extension(it)),
+    };
+    iter::once(path)
+        .chain(with_extension)
+        .find(|it| it.is_file())
+}
+
+/// [`lookup_the_nearest_file_dir`] will start from [`from`] and search for file [`the_nearest_file`] in the parent directories.
+/// If found, it will return the [`Some`] of [`the_nearest_file`] file path. If not found, it will return [`None`]
+pub(crate) fn lookup_the_nearest_file_dir(
```
Comment (zong-zhe) [inline kclvm/driver/src/kpm_update.rs:79]: The same as the method kcl(), you can move it to https://github.com/kcl-lang/kcl/blob/main/kclvm/driver/src/lib.rs

File: kclvm/driver/src/kpm_update.rs [hunk 5]
```diff
@@ -0,0 +1,96 @@
+use anyhow::{bail, Result};
+use std::{collections::HashMap, env, iter, path::PathBuf, process::Command};
+
+const MANIFEST_FILE: &str = "kcl.mod";
+
+/// Update the KCL module.
+///
+/// This function calls `kcl mod update` to update the KCL module.
+pub(crate) fn update_kcl_module(manifest_path: PathBuf) -> Result<()> {
```
Comment (zong-zhe) [inline kclvm/driver/src/kpm_update.rs:9]: You can try to add some test cases for `update_kcl_module()`

File: kclvm/driver/src/kpm_update.rs (hunk 1 above)
Comment (octonawish-akcodes) [reply kclvm/driver/src/kpm_update.rs]: okay

File: kclvm/driver/src/kpm_update.rs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    }
+}
+
+/// Get the path for the KCL executable.
+fn kcl() -> PathBuf {
+    get_path_for_executable("kcl")
+}
+
+/// [`get_path_for_executable`] will return the path for [`executable_name`].
+pub fn get_path_for_executable(executable_name: &'static str) -> PathBuf {
```
[the remark this answers is not in the document]
Comment (octonawish-akcodes) [reply kclvm/driver/src/kpm_update.rs]: okay

File: kclvm/driver/src/kpm_update.rs (hunk 2 above)
Comment (octonawish-akcodes) [reply kclvm/driver/src/kpm_update.rs]: okay

File: kclvm/driver/src/kpm_update.rs (hunk 4 above)
Comment (octonawish-akcodes) [reply kclvm/driver/src/kpm_update.rs]: okay

File: kclvm/driver/src/kpm_update.rs (hunk 5 above)
Comment (octonawish-akcodes) [reply kclvm/driver/src/kpm_update.rs:10]: I tried to add basic functions for test cases for the `update_kcl_module`. While I was exploring the tests.rs file I got some confusions as follows:

1. What is this `test_vendor` file for?
2. Why do we have  expand file pattern dir?
3. I am bit confused about the path passing to kcl mod update from the metadata function that we already have?

I was following the kpm_metadata test cases implementation, also wanted to know about the use case of this `test_fill_pkg_maps_for_k_file` and `test_fill_pkg_maps_for_k_file_with_line`?

cc @zong-zhe

Comment (zong-zhe) [reply kclvm/driver/src/kpm_update.rs:10]: - `test_vendor` is a tmp dir for test, used to test whether the environment variable `$KCL_PKG_PATH` is working properly. This environment variable is used to specify the directory in which the KCL dependencies is stored.

- `test_fill_pkg_maps_for_k_file ` is used to test the function `fill_pkg_maps_for_k_file`, which will find the nearest   `kcl.mod` file from the *.k file, and get the dependency path from the `kcl.mod`.

- `test_fill_pkg_maps_for_k_file_with_line` and `test_fill_pkg_maps_for_k_file` is the similar, the difference is that `test_fill_pkg_maps_for_k_file_with_line` needs to test the special processing of underscores and horizontal lines in the name of the dependencies.

**Comment by octonawish-akcodes** [comment]
I added the test cases too @Peefy @zong-zhe

**Comment by Peefy** [comment]
> I added the passing test cases too @Peefy @zong-zhe

PR conflict, can you rebase the main branch and refactor it?

**Comment by octonawish-akcodes** [comment]
@Peefy check now?

File: kclvm/driver/src/tests.rs [hunk 7]
```diff
@@ -382,28 +383,42 @@ fn test_get_pkg_list() {
...
-            )),
-        }
-    }
-    let result = update_kcl_module(manifest_path);
-    assert!(result.is_err());
-}
+fn test_update_module() {
+    test_fetch_mod_metadata();
+    println!("test_update_module() passed");
+}
```
Comment (Peefy) [inline kclvm/driver/src/tests.rs:424]: Pls format the code

Comment (octonawish-akcodes) [reply kclvm/driver/src/tests.rs]: done

**Comment by Peefy** [comment]
> @Peefy check now?

Thank you! Please sign the DCO

File: kclvm/driver/src/tests.rs [hunk 8]
```diff
@@ -382,28 +383,42 @@ fn test_get_pkg_list() {
...
-        MockCommand {
-            output: Err(std::io::Error::new(
-                std::io::ErrorKind::Other,
-                "Command failed",
-            )),
-        }
-    }
-    let result = update_kcl_module(manifest_path);
-    assert!(result.is_err());
+fn test_update_module() {
```
Comment (Peefy) [inline kclvm/driver/src/tests.rs:421]: Hi @octonawish-akcodes  CI failed and would you please sign the DCO?

Comment (octonawish-akcodes) [reply kclvm/driver/src/tests.rs:421]: im inspecting that, and i am having trouble signing dco due to conflicts :/

**Pull request closed without merging**